### ISSUE TITLE
NoSQL Support

### ISSUE THREAD

**Issue by casoetan** [qor/admin#113]
Is it possible to use with a no sql db like MongoDB. I noticed it is tightly coupled with gorm which dos not support this.

**Comment by gstvg** [comment]
You need to overwrite the CRUD handlers. Take a look https://doc.getqor.com/admin/extend_admin.html#overwrite-curd-handler . The default QOR handlers is written in only ~130 lines. Looks easy to rewrite them to another storage engine. But be aware that by now is not possible to rewrite Singleton resources handlers.

**Comment by mrclayman** [comment]
@gstvg , could you elaborate on that approach a little more? I understand the idea of overriding the handlers, but how do I leverage resource loading through these handlers in my view controllers that I register into routers? When I look at the QOR example site, I see code like this:
	var (
		Orders      []orders.Order
		currentUser = utils.GetCurrentUser(req)
		tx          = utils.GetDB(req)
	)`

        tx.Preload("OrderItems").Where("state <> ? AND state != ?", orders.DraftState, "").Where(&orders.Order{UserID: &currentUser.ID}).Find(&Orders)
`

The general idea of this code is clear to me, but it still uses gorm.DB internally to load the resource. How can I make QOR use the handlers associated with the resource?

**Comment by cryptix** [comment]
@mrclayman: you are mixing a gorm example with the qor data access interface.

If you want to use any other db in your application (in qor-example this is the shop) you use that interface just like in any other.

If you want to use the /admin interface with anything other than gorm, that’s when you replace the handlers. It doesn’t magically retrofit gorm with querying abilities.

**Comment by mrclayman** [comment]
Thank you very much for your response, @cryptix . I guess I am mixing things up, I am totally new to qor. Since I am only interested in the admin component of qor, I did override the 4 handlers and was able to read data from my MongoDB backend and display them in a table. However, I assume I am missing some additional routes because when I click on any item in that table, another window is shown with the exact same table. I guess I am going to need to further customize the behavior of the resource.

Btw. would you mind popping up on Gitter? I am there as mrclayman and I would appreciate it very much if you could provide me with a little more information.

**Comment by sachin-puranik** [comment]
As I have understood till now, Just by replacing these 4 functions, we would be able to change the data storage and retrieval. But Ideally shouldn't it be done by porting QOR itself with such data base.
Is there any possibility to use the GORM with say google DataStore which is a flat database.
How would db.AutoMigrate() work...

**Comment by cryptix** [comment]
>  I guess only after porting the dialect.

yup - that's the alternative.

make something that gorm can use and dialects is the way to do just that.

**Comment by sachin-puranik** [comment]
While I was reading yesterday I found that dialect + basic sql interface is
what it needs .... and execute function do not know the internal details of
query ....

But in case of flats database , it needs internal details of DB object ...
and use that in different way than SQL ... so in such case SQLcore object
should be part of form so that it has access to form internal objects and
can make use of them ...

How do u think the approach would be ?

**Comment by sachin-puranik** [comment]
Typo error ... plz read form and GORM

**Comment by cryptix** [comment]
please open a new issue about this at the gorm issue tracker. this is not the place.

**Comment by sachin-puranik** [comment]
Sure. Thanks. I thought it was like mailing list.
I have raised the issue on GORM. Will add few more comments.

**Comment by WoolWarrior** [comment]
I am working on qor-admin with a NoSQL database. 

I started with the example in document: https://doc.getqor.com/get_started.html
Code using sqlite3 (exactly as the example above): https://gist.github.com/WoolWarrior/2bd7ffe68d1ddefa2862bbd4723ab535

I tried to overwrite the CRUD handler as the doc suggested:
https://doc.getqor.com/admin/extend_admin.html#overwrite-curd-handler
 
My initial code not working as expected. Basically, an assignment does not work properly: 
`result = &product101`
1st attempt code:
https://gist.github.com/WoolWarrior/90eec2d82ee6fa6201ea108db97f6b41

I spent 3 days to figure the solution. Finally, used deep copy concept to replace naive assignment:
either https://godoc.org/github.com/getlantern/deepcopy#pkg-files
or https://www.jianshu.com/p/f1cdb1bc1b74
2nd attempt code (seems working well):
https://gist.github.com/WoolWarrior/3b95a0c0579af53c040c80081a7e7c98

**Comment by WoolWarrior** [comment]
Finally managed to get QOR-admin working with NoSQL (AWS DynamoDB)

overwrite CRUD handler in func ConfigureQorResource and func ConfigureQorResourceDynamoDB
https://gist.github.com/WoolWarrior/add5a8adab34d4b8522be7008ba93cdb

**Comment by sachin-puranik** [comment]
How can the other sections work with  aws , google data store or Dynamo ,
since GORM is used ...
Can gorm be also used with these databases ....


Thanks
